### ISSUE TITLE
Feature request: AceStep 1.5 XL audio LM context (lm_hints) for quality parity with ComfyUI

### ISSUE THREAD

**Issue by MarcBate** [ostris/ai-toolkit#921]
[before.mp3](https://github.com/user-attachments/files/29448639/before.mp3)
[after.mp3](https://github.com/user-attachments/files/29448640/after.mp3)

## Summary

AceStep 1.5 XL training and inference in ai-toolkit currently produces significantly lower quality output than ComfyUI. The root cause is that ai-toolkit feeds **silence as context latents** on every denoising step, while ComfyUI generates proper **audio semantic codes** (`lm_hints`) via a Qwen3 LM pipeline.

## What ComfyUI does (that ai-toolkit is missing)

The AceStep 1.5 XL DiT expects a 128-dim context tensor per frame: the first 64 channels carry `lm_hints` (audio semantic content), the last 64 are `chunk_masks` (all ones). When silence (zeros) is used instead of `lm_hints`, the model is deprived of its primary audio-semantic guidance signal.

ComfyUI's pipeline:

1. **Qwen3 ACE15 LM** (`qwen_4b_ace15.safetensors` or `qwen_1.7b_ace15.safetensors`) generates audio semantic tokens autoregressively — `duration × 5` tokens at 5 Hz, sampled from vocab range `[151669, 215669]` (64000 audio tokens)
2. **FSQ quantizer** (`ResidualFSQ`, levels=`[8,8,8,5,5,5]`, codebook size=64000, fsq_dim=2048) decodes token indices → continuous embeddings via `get_output_from_indices`
3. **AudioTokenDetokenizer** upsamples 5 Hz → 25 Hz (pool_window_size=5), outputs `[B, T_25Hz, 64]`
4. `context_latents = cat([lm_hints, chunk_masks], dim=-1)` → `[B, T, 128]` fed to DiT each step

All three components live in the AIO base model safetensors under `model.diffusion_model.tokenizer.*` and `model.diffusion_model.detokenizer.*` — they're already in the file, just not loaded by ai-toolkit.

## What I implemented (in my fork)

I ported the full pipeline in two new files with no ComfyUI dependencies:

- `audio_lm.py` — pure-PyTorch Qwen3 4B/2B LM with KV-cached autoregressive generation, CFG, top-p/k/min-p filtering
- `audio_tokenizer.py` — `ResidualFSQ` + `AudioTokenDetokenizer`, weights extracted directly from the AIO safetensors, all params inferred from weight shapes

**Result:** quality is dramatically better — the model produces coherent, musically structured output instead of vague noise. LoRA fine-tuning also benefits since training and inference now use matching context distributions.

## Key bug to be aware of

During porting, I hit a subtle KV cache bug: if you pre-allocate KV cache tensors before the initial prompt pass (for efficiency), the causal mask condition `is_causal = (T > 1 and kv_cache is None)` evaluates to `False` even during full-prompt processing — because the cache is pre-allocated, not None. This means the prompt forward pass runs **without causal masking**, corrupting the LM's internal state and producing garbage audio codes (audible as noise). Fix: `is_causal = (T > 1)` regardless of cache state.

## Suggested implementation for ai-toolkit

The cleanest integration would use `transformers.AutoModelForCausalLM` to load the Qwen3 model (avoiding the need to hand-roll the architecture), plus a small `ResidualFSQ` and `AudioTokenDetokenizer` that match the AIO key layout exactly:

```python
# In model config
model:
  name_or_path: ostris/ace_step_1.5_ComfyUI_files/ace_step_1.5_xl_base_aio.safetensors
  audio_lm_path: /path/to/qwen_4b_ace15.safetensors  # or 1.7b variant

# In pipeline.__call__
if context_latents is not None:
    ctx = context_latents
else:
    # existing silence fallback
    sil = get_silence_latent(latent_len, device, dtype)
    ctx = cat([sil.T, ones_like(sil.T)], dim=-1)
```

The tokenizer/detokenizer weights are already in the AIO file — no extra download needed for those. Only the Qwen3 LM weights need a separate file.

## Notes

- Works with the XL AIO format (`ostris/ace_step_1.5_ComfyUI_files`); non-XL untested but should work since all params are inferred from shapes
- The Qwen3 4B model is 7.8 GB — for GPU memory constrained setups, generate codes first (LM → CPU), then run diffusion
- Negative prompt / CFG for LM: use `cfg_scale=2.0` with positive (metadata YAML in `<think>` block) and negative (empty `<think>`) prompts
- The 1.7B model works but produces slightly less coherent codes than 4B

Happy to share the full implementation if it helps.
